### PR TITLE
i2c-mcux-lpi2c: take semaphore during transfer

### PR DESCRIPTION
**Pull request by pdgendt**
Ported commit 49a3ce5881227f63b95dd2bef153a391161d06b8 for a very similar driver, where transfer locking is required if multiple devices use the same i2c bus.

### CODE REVIEW

**Comment by GunZi200** [comment]
Could you please add handling for the return value of `k_sem_take()` and `k_sem_init()` to avoid generating a coverity report in the future. 

If the return value is not needed, then add a `(void) `cast like I did here: https://github.com/zephyrproject-rtos/zephyr/pull/33223 This will prevent a false coverity report. 😃

**Comment by pdgendt** [comment]
I've added the check for `k_sem_take` however nowhere in the zephyr code base is this added for `k_sem_init`.

**Comment by GunZi200** [comment]
> I've added the check for `k_sem_take` however nowhere in the zephyr code base is this added for `k_sem_init`.

I just saw that it had a return value here https://docs.zephyrproject.org/apidoc/latest/group__semaphore__apis_gadcd0e6cfba3392fb887222eafe4c1845.html

Maybe the return value hasn't been checked since it's never been a problem. It seems to me that if it doesn't return `0` then the parameters input into  `k_sem_init()` are invalid and thus it returns `-EINVAL`

----

Now I've got an idea for why Coverity has not created a report for `k_sem_init()` before...

If I look at the report it made on `k_sem_take()` https://github.com/zephyrproject-rtos/zephyr/issues/33034, it says: 
> Calling "k_sem_take" without checking return value (as is done elsewhere 13 out of 15 times).

Note: `as is done elsewhere 13 out of 15 times`. So now I'm wondering if the return value of `k_sem_init()` is never checked anywhere in the code base. Then Coverity doesn't know this is an issue? 🤔 

A quick google search shows this link: https://community.synopsys.com/s/article/Coverity-doesn-t-report-a-unchecked-return-value

> This is a statistical checker. As described in the Checker Reference document, part of how the checker determines which functions should be checked is by statistically analyzing usage patterns across the code base.

I don't have access to Zephyr's verson of "Checker Reference" `https://<your_coverity_url>/doc/en/cov_checker_ref.html`

I'm not a reviewer btw, just find this interesting :)

@MaureenHelm what are your thoughts on this?

File: drivers/i2c/i2c_mcux_lpi2c.c [hunk 1]
```diff
@@ -240,17 +261,17 @@ static const struct i2c_driver_api mcux_lpi2c_driver_api = {
 	static struct mcux_lpi2c_data mcux_lpi2c_data_##n;		\
 									\
 	DEVICE_DT_INST_DEFINE(n, &mcux_lpi2c_init, device_pm_control_nop,\
-			    &mcux_lpi2c_data_##n,			\
-			    &mcux_lpi2c_config_##n, POST_KERNEL,	\
-			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
-			    &mcux_lpi2c_driver_api);			\
+				&mcux_lpi2c_data_##n,			\
```
Comment (pabigot) [inline drivers/i2c/i2c_mcux_lpi2c.c:264]: In general please try to avoid unrelated whitespace changes in commits.  No need to back this one out.

**Review by pabigot — CHANGES_REQUESTED** [review]
Please replace the commit message body with a self-contained description of the intent behind the change.  Even just "Transfer locking is required if multiple devices use the same i2c bus." from the PR description would cover it.

File: drivers/i2c/i2c_mcux_lpi2c.c (hunk 1 above)
Comment (pdgendt) [reply drivers/i2c/i2c_mcux_lpi2c.c]: Removed the whitespace changes.

**Review by pabigot — APPROVED** [review]

**Comment by galak** [comment]
Can we add anything to tests/driver/i2c/i2c_api to attempt to test the issue of different threads accessing the same i2c controller?

**Comment by pabigot** [comment]
> Can we add anything to tests/driver/i2c/i2c_api to attempt to test the issue of different threads accessing the same i2c controller?

I don't think that should be an issue for this PR.

Outside this PR I don't see how this can't be done generally.  Individual I2C drivers that have to implement the mutex capability under the current device mutex architecture, so using the I2C emulator won't get us very far.  To test individual drivers we'd need devices that have two I2C peripherals on the bus, and a way to at least semi-reliably detect that simultaneous interaction doesn't fail.

**Comment by galak** [comment]
> > 
> 
> I don't think that should be an issue for this PR.
> 
agreed, was just a question raised by reviewing this PR.

> Outside this PR I don't see how this can't be done generally. Individual I2C drivers that have to implement the mutex capability under the current device mutex architecture, so using the I2C emulator won't get us very far. To test individual drivers we'd need devices that have two I2C peripherals on the bus, and a way to at least semi-reliably detect that simultaneous interaction doesn't fail.

**Review by galak — APPROVED** [review]
